### ISSUE TITLE
Possibility to Paste copied contours from external app (Illustrator / RoboFont / Glyphs)

### ISSUE THREAD

**Issue by BlackFoundry** [googlefonts/fontra#260]
Paste the clipboard content as contours if possible in a Fontra glyph

**Comment by nedkamburov** [comment]
The Clipboard API supports a few options. The one that might interest us is 'copying images' which is disabled in Firefox by default. It needs to be enabled. In other browsers it just works.

A good resource to explain a little bit about different MIME Types (part of `ClipboardItem` object)
https://www.stefanjudis.com/notes/a-clipboard-magic-trick-how-to-use-different-mime-types-with-the-clipboard/
and 
https://web.dev/async-clipboard/

Testing is done by monitoring **Finder->Edit->Show Clipboard**. Later we can see how it works (IF it works) in this sketch:
https://codepen.io/nedkamburov/pen/BaPemzN 

— plain text from anywhere works fine. Formatting can also be copied ( in RTF )
— Robofont outputs simple XML
— Glyphs2 outputs the actual shape I want to copy somehow. Glyphs3 — nothing
— Illustrator 2020 outputs an svg

TLDR; 
If all programs provide text representation of the shapes/paths, we can use the clipboard API in all browsers. However, we will have to convert all of them to something we can use, such as JSON.

In the browser you can copy: 
— plain text
— Robofont path
— Illustrator path
— Glyphs (not working?)

**Comment by justvanrossum** [comment]
It would be cool if the codepen could list all the mime types it finds, even without showing the contents.

**Comment by justvanrossum** [comment]
(I would ismplify the codepen by using a stupid "paste" button, or "display clipboard" button instead of the command-v stuff. We're just peeking at the clipboard data after all.)

**Comment by justvanrossum** [comment]
DrawBot toy to inspect clipboard contents on macOS:
```python
from AppKit import NSPasteboard

pb = NSPasteboard.generalPasteboard()

for item in pb.pasteboardItems():
    print("----")
    for tp in item.types():
        print(tp)
        # if "svg" in tp:
        #     print("--", item.stringForType_(tp))
        #     # print("--", item.dataForType_(tp))
        #     # print("--", item.propertyListForType_(tp))
```

https://developer.apple.com/documentation/appkit/nspasteboarditem?language=objc

**Comment by nedkamburov** [comment]
https://codepen.io/nedkamburov/pen/BaPemzN  
Okay, so I integrated your suggestions. Now that demo shows:
— All available types
— All outputs (if the you paste a PNG, it will show you a [blob](blob:https://cdpn.io/f824d1bb-ef5d-4d99-b7f6-87e1f7428af0) which you can paste in another tab and view.

**Comment by nedkamburov** [comment]
Quick experiments shows that RoboFont spits out content in those types:
```
dyn.ah62d4rv4gu81e55cr7xg85xyk70hw6dknrw023pdsvy085wuqf31kuxtqf3gkzd3sbwu
com.adobe.pdf
public.utf8-plain-text
```

Glyphs has those:
```
dyn.ah62d4rv4gu8ys5d3sbyhgkdfrvw043pssv3wa6dbsr4gn2xtqf3gkkdytf2gnkd0gq
dyn.ah62d4rv4gu8ys5d3sbyhgkdfrvw043pssv3wa6dbsr4gn2xtqf3gkkdytf2gn
com.adobe.pdf
dyn.ah62d4rv4gk8ycwndkk
com.adobe.encapsulated-postscript
```
If I am not mistaken AI also has `com.adobe.pdf` so this one seems to be the most universal.

I cannot seem to find which formats are fully supported among all browsers. What I see **text/plain, text/html and image/png** are widely supported.

However, there are custom formats that Chromium (Chrome/Safari/Edge) can use https://developer.chrome.com/blog/web-custom-formats-for-the-async-clipboard-api/ I already added it to the sketch so you can see that it works. It outputs it as `web image/svg+xml` at the moment. In essence, in can be anything.

----
I don't know how to proceed from here. As far as I can see we can export one Blob in several formats but it really depends what the other platforms expect. Robofont is using an SVG but inside it has a glyph attribute, Glyphs works with adobe.pdf. So what experiments should take place next?
Second thing: What are we going to do with Firefox if it has so many limitation for the Clipboard API?

**Comment by justvanrossum** [comment]
Thanks. Of all the formats we see in native macOS, it seems that only `text/plain` arrives in the browser, which may be SVG (from Ai) or GLIF (from RoboFont, it's XML but not SVG). 

From Glyphs nothing arrives: we may have to open a feature request for Glyphs to put SVG on the clipboard, possible also as plain text.

Does Glyphs accept SVG for paste?

PDF is mostly useless for us as it's super hard to parse. PNG is useless for us as it's pixels.

Firefox: if people explicitly have to allow it, than so be it.

Is it possible to put a custom/private type on the clipboard within the browser, and read it back within the same browser? It would be cool if we could do that for internal copy/paste, so we can use the plain text part for SVG or GLIF.

**Comment by nedkamburov** [comment]
> Is it possible to put a custom/private type on the clipboard within the browser, and read it back within the same browser? It would be cool if we could do that for internal copy/paste, so we can use the plain text part for SVG or GLIF.

That's what i meant with this:
> However, there are custom formats that Chromium (Chrome/Safari/Edge) can use https://developer.chrome.com/blog/web-custom-formats-for-the-async-clipboard-api/ I already added it to the sketch so you can see that it works. It outputs it as web image/svg+xml at the moment. In essence, in can be anything.

It works well, I think. You get the content and the format is recognised, so yes.

**Comment by justvanrossum** [comment]
> In essence, in can be anything

Ah yes, that's cool. Does this "can be anything" work in other browsers? (Again, this would only be for copy/paste-within-Fontra-itself)

**Comment by nedkamburov** [comment]
Firefox is the culprit. Even with those special settings on. It still complains
``` Uncaught (in promise) DOMException: Type 'web image/svg+xml' not supported for write ```
I will check it again.

**Comment by nedkamburov** [comment]
Todos:
[x] Copy SVG from Fontra to other applications (Illustrator, Figma, Glyphs?)
[] Copy Glyph from Fontra to other applications (RoboFont)
[] Copy GLIF from Fontra to other applications

[] Paste SVG into Fontra
[] Paste Glyph into Fontra  (RoboFont)
[] Paste GLIF into Fontra
